### PR TITLE
#6061 generating superpom structure

### PR DESCRIPTION
**Pull request by Tcharl**
- Please make sure the below checklist is followed for Pull Requests.
This PR should work, but, jhipster guys shall: 
* clone the https://github.com/Tcharl/jhipster-parent
* release it (changing the scm url in the parent pom, then following the readme.

After that, the generator-jhipster tests and generation should be all green :)

Enjoy !
- [ ] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
- [x] Tests are added where necessary
- [x ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed (I think so, may be some indentations to review later)

### CODE REVIEW

**Comment by deepu105** [comment]
@Tcharl are you still on this?

**Comment by deepu105** [comment]
@jhipster/developers guys can someone review this

**Comment by deepu105** [comment]
@atomfrede do you think this affects gradle build setup?

**Comment by atomfrede** [comment]
Not sure really. Maybe we need an own jhipster plugin like the spring boot plugin. But I need to take a look at the documentation cause that's usually not a setup I use with gradle.

**Comment by Tcharl** [comment]
Hi,
Sorry for the late answer, I didn't see the comments (even if following jh PR on a day to day basis)
I rebased it, it should be ok. 
I don't think it would affect any gradle stuff as generators are different.

Note that in order to have it working, the jhipster community shall for and release [jhipster-parent](https://github.com/Tcharl/jhipster-parent) bom (like spring/infinispan/whatever framework does).

Best regards,
Charlie

**Comment by PierreBesson** [comment]
@Tcharl I'm not sure the team is 100% certain that we should migrate to our own parent. This probably makes a lot of sense but it has also the big drawback that it ties our users very strongly to the JHipster project. I think many users wouldn't mind to have this as it eases upgrades but it also adds complexity for contributors because they will need to understand the pom structure.
Initially JHipster was not created with the ambition of being a full fledged framework. The original idea was that a JHipster app is similar to what you could have coded yourself, we are just saving time for the developer by integrating the building blocks for them.

So I think we should have a discussion in person, there should be a JHipster meetup in Paris in the coming weeks. Maybe we can see you there ?

**Comment by Tcharl** [comment]
Hi @PierreBesson,

I've not implemented a parent pom, just a bom (containing only dependencyManagement).

Please, take a look at the [corresponding issue](https://github.com/jhipster/generator-jhipster/issues/6061) (and please answer on it, because usually, the team use tickets for correspondance: reviews comments are only concerning discussion around the code).

From what I know, this solution was proposed by @erikkemperman and it has been a long discussion thread to choose the right way to do.

Regards,

**Comment by atomfrede** [comment]
We could use the bom also for dependency management in gradle +1

**Comment by erikkemperman** [comment]
Actually my idea was to have a parent not as a separately released artifact, just a separate pom within the generated directories to keep the main user-visible pom clean and avoid conflicts when upgrading.

But the dependencies should not be in there, instead in a separately released `jhipster-dpendencies` (so the gradle plugin can use them).

Sorry I haven't had any time at all to work on this stuff... Really busy with other things at work, and my GF moved in so lots of redecorating and construction there :-)

**Comment by erikkemperman** [comment]
@Tcharl Aargh, I did not realize you were still working on this... You did not answer when I [asked](https://github.com/jhipster/jhipster-dependencies/pull/2#issue-263458921) (I think?) and there was no activitiy on this thread for a long while so I assumed not and submitted my own PR to use the new BOM, https://github.com/jhipster/generator-jhipster/pull/6509.

I apologize for the confusion.

Interestingly, there seem to be quite some differences between our approaches. Of course I just wanted to add the BOM, whereas you appear to still want to add the several super-POMs, as well as those "steroids" all in a single PR? Also, no Gradle?

**Comment by Tcharl** [comment]
Hi @erikkemperman, from my side too, I wasn't too meticulous: I think that tour PR only modified the Gradle part (I just looked at the upper part of the commits while reviewing :-S).
This work is pretty hard to follow, as we have many tickets for the issue: I thought that as the 6131 mentionned that there will be two PRs (one without steroids, and an other bringing them), we can continue to work in parallel on it.
So I think that there is NP at all: #6509 will be about bringing the Bom to Maven and Gradle (I was just focused on Maven on my side, as I thought that atomfred was the Gradle guy) and this one about steroids. Not sure that one will be accepted, but at least it will be useful within my company projects: 
* The need of a maven site and license header generation is mandatory as we're doing Inner Source : so every project should have a dedicated site listing team/org, with their COTS license, have an overview of dependency convergence (I'm working in a security critical company, so this information is important to identify holes), and showing the overhall 'maturity' of the project (to let the consumer choose between use/make).
* The use of maven release plugin is really a plus, as most of our developers aren't familliar with a release process (and we've ARMs everywhere), and aren't so aware of it (for example, creating a tag for a release is not an automatism, and sometimes it become problematic for patch release on old products).
* And finally, I would prefer having failsafe for Integration tests, as I'm kind of a TDD guy, and my Infinitest builds began to slow my devs :-P.

At least, it could become a module once your PR will pass as the maintenance of the pom will be lighter (no need to update parts of the pom at each dependency upgrade).
When do you think your PR will be ready? if it's in weeks, I'll rebase this PR on your branch, as it can be simpler in term of dev.

Best regards, and good luck for that bom integration!

**Comment by erikkemperman** [comment]
@Tcharl I'm not sure what the estimated time is on https://github.com/jhipster/generator-jhipster/pull/6509 -- I think it's done from my side, just the occasional rebase when master changes so much it conflicts. At first Julien suggested it might be soon after 4.10, but that's now been released for a little while, but Deepu labeled it "major" the other day, which I'm not sure what it means :-)
